README.md
# Logv 文档

## 概述

logv is logger of vdms project for golang.

logv didn't impelement its own interface, it neseted logrus.Logger, so all logrus logger interface 
is avaliable. For example:
Warn() Debugln() Errorf()

## 使用方法

### 方法一

    调用logv.New()， 创建logv.Logger实例，使用该实例

### 方法二

    调用logv.Errorf()或logv.Debugf()等接口（推荐此方法）,使用SetLogger函数可以设置全局logger属性

    注意：
    1. SetRotate函数必须在SetOutputFile之后调用
    2. SetOutput会移除默认的日志回滚


ChangeLog.md
# Change Log

## version: v1.0.0

版本更新时间: 2019.12.30 11:43

功能: 封装 logrus 和 lumberjack日志库

## version: v1.0.1

版本更新时间: 2019.12.30 14:14

当前功能:

- 添加全局日志函数及默认全局 logger
- 添加配置全局 logger 的 SetLogger 函数


logv_test.go
package logv

import "testing"

func TestNew(t *testing.T) {
	f := &Format{
		Filename:   "./a.log",
		MaxSize:    1,
		MaxAge:     1,
		MaxBackups: 3,
		Loglevel:   DebugLevel,
	}
	logger := New(f)
	for i := 0; i < 100000; i++ {
		logger.Warnln(i)
		logger.Traceln(i)
		logger.Debugln(i)
		logger.Errorln(i)
		logger.Infoln(i)
	}
}


logger_test.go
package logv

import (
	"os"
	"testing"
)

func TestLog(t *testing.T) {
	Tracef("trace %s", "trace")
	Debugf("Debugf %s", "Debugf")
	Infof("Infof %s", "Infof")
	Warnf("Warnf %s", "Warnf")
	Errorf("Errorf %s", "Errorf")
}

func TestSetOutputFile(t *testing.T) {
	SetOutputFile("./a.log")
	Tracef("trace %s", "trace")
}

func TestSetRotate(t *testing.T) {
	SetRotate(1, 2, 1)
	for i := 0; i < 100000; i++ {
		Debugf("%d", i)
	}
}

func TestSetOutput(t *testing.T) {
	f, err := os.Create("./b.log")
	if err != nil {
		t.Errorf("create b.log failed")
	}
	defer f.Close()
	SetOutput(f)
	Tracef("trace %s", "trace")
}

// func TestFatalf(t *testing.T) {
// 	Fatalf("Fatalf %s", "Fatalf")
// }

// func TestPanicf(t *testing.T) {
// 	Panicf("Panicf %s", "Panicf")
// }

func TestSetLogger(t *testing.T) {
	SetLogger(&Format{
		Filename:   "./b.log",
		MaxSize:    1,
		MaxAge:     1,
		MaxBackups: 4,
		Loglevel:   InfoLevel,
	})
	Tracef("trace %s", "trace")
	Debugf("Debugf %s", "Debugf")
	Infof("Infof %s", "Infof")
	Warnf("Warnf %s", "Warnf")
	Errorf("Errorf %s", "Errorf")
}


logv.go
package logv

import (
	"io"
	"os"

	"github.com/sirupsen/logrus"
	"gopkg.in/natefinch/lumberjack.v2"
)

// Logger is log type for package logv.
type Logger struct {
	// logrus.Logger is underline logger.
	*logrus.Logger

	// lumberjackLogger is used to rotate log file, lumberjackLogger is not set by default,
	// this will be set when SetOutput function is being called
	lumberjackLogger *lumberjack.Logger

	filename string
}

// Format used to set up log file format
type Format struct {
	Filename string
	// MaxSize uses megabytes
	MaxSize int
	// MaxBackups set backup log file number
	MaxBackups int
	// MaxAge uses days, MaxAge set the lifetime of backup log file
	MaxAge int
	// Loglevel
	Loglevel Level
}

const (
	defaultMaxSize    = 50
	defaultFilename   = "./logv.log"
	defaultMaxBackups = 10
	defaultMaxAge     = 10
)

// Level is log level type
type Level uint32

// logging levels
const (
	UnknownLevel Level = iota
	PanicLevel
	FatalLevel
	ErrorLevel
	WarnLevel
	InfoLevel
	DebugLevel
	TraceLevel
)

// NewDefault create new logger that output to stdout.
// No rotate. Default logrus logger.
func NewDefault() *Logger {
	return &Logger{
		Logger: logrus.New(),
	}
}

// New create new logger that output to stdout.
// Default Value:
//         	Filename: ./logv.log
//			MaxSize: 50
//			MaxBackups: 3
//			MaxAge: 10
//			Loglevel: PanicLevel
func New(f *Format) *Logger {
	if len(f.Filename) == 0 {
		f.Filename = defaultFilename
	}
	if f.MaxSize <= 0 {
		f.MaxSize = defaultMaxSize
	}
	if f.MaxBackups <= 0 {
		f.MaxBackups = defaultMaxBackups
	}
	if f.MaxAge <= 0 {
		f.MaxAge = defaultMaxAge
	}

	lumberjackLogger := &lumberjack.Logger{
		Filename:   f.Filename,
		MaxSize:    f.MaxSize,
		MaxBackups: f.MaxBackups,
		MaxAge:     f.MaxAge,
	}
	multiIO := io.MultiWriter(os.Stdout, lumberjackLogger)

	l := logrus.New()
	l.SetOutput(multiIO)
	l.SetReportCaller(true)
	l.SetFormatter(&logrus.TextFormatter{
		DisableColors:   false,
		FullTimestamp:   true,
		TimestampFormat: "2006-01-02 15:04:05",
		ForceColors:     true,
	})
	// Only log the warning severity or above.
	l.SetLevel(logrus.Level(transLevelToLogrusLevel(f.Loglevel)))

	return &Logger{
		Logger:           l,
		lumberjackLogger: lumberjackLogger,
		filename:         f.Filename,
	}
}

func transLevelToLogrusLevel(l Level) logrus.Level {
	var logrusLevel logrus.Level
	switch l {
	case PanicLevel:
		logrusLevel = logrus.PanicLevel
	case FatalLevel:
		logrusLevel = logrus.FatalLevel
	case ErrorLevel:
		logrusLevel = logrus.ErrorLevel
	case WarnLevel:
		logrusLevel = logrus.WarnLevel
	case InfoLevel:
		logrusLevel = logrus.InfoLevel
	case DebugLevel:
		logrusLevel = logrus.DebugLevel
	case TraceLevel:
		logrusLevel = logrus.TraceLevel
	default:
		logrusLevel = logrus.InfoLevel
	}
	return logrusLevel
}


logger.go
package logv

import (
	"io"
	"os"

	"github.com/sirupsen/logrus"
	"gopkg.in/natefinch/lumberjack.v2"
)

var logger *Logger

func init() {
	logger = NewDefault()
	logger.SetLevel(logrus.TraceLevel)
	logger.SetFormatter(&logrus.TextFormatter{
		DisableColors:   false,
		FullTimestamp:   true,
		TimestampFormat: "2006-01-02 15:04:05",
		ForceColors:     true,
	})
}

// SetLogger set default global logger config
func SetLogger(f *Format) {
	var filename string
	var maxSize, maxBackups, maxAge int
	var l Level

	if f.Filename == "" {
		filename = defaultFilename
	} else {
		filename = f.Filename
	}

	if f.MaxSize == 0 {
		maxSize = defaultMaxSize
	} else {
		maxSize = f.MaxSize
	}

	if f.MaxBackups == 0 {
		maxBackups = defaultMaxBackups
	} else {
		maxBackups = f.MaxBackups
	}

	if f.MaxAge == 0 {
		maxAge = defaultMaxAge
	} else {
		maxAge = f.MaxAge
	}

	lumberjackLogger := &lumberjack.Logger{
		Filename:   filename,
		MaxSize:    maxSize,
		MaxBackups: maxBackups,
		MaxAge:     maxAge,
	}
	multiIO := io.MultiWriter(os.Stdout, lumberjackLogger)
	logger.SetOutput(multiIO)
	logger.lumberjackLogger = lumberjackLogger
	logger.filename = filename

	logger.SetLevel(transLevelToLogrusLevel(l))
}

// SetOutputFile set default logger output file, if rotate is configed, reuse configuration,
// otherwise config use default value,
// 		MaxSize: 50 mega bytes
//		MaxBackups: 10
// 		MaxAge: 10 days
func SetOutputFile(filename string) {
	if logger.lumberjackLogger == nil {
		lumberjackLogger := &lumberjack.Logger{
			Filename:   filename,
			MaxSize:    defaultMaxSize,
			MaxBackups: defaultMaxBackups,
			MaxAge:     defaultMaxAge,
		}
		multiIO := io.MultiWriter(os.Stdout, lumberjackLogger)
		logger.SetOutput(multiIO)
		logger.lumberjackLogger = lumberjackLogger
		logger.filename = filename
	}
}

// SetRotate set log rotate info, if filename is configed, reuse configuration,
// otherwise config use default value,
// 		Filename: ./logv.log
func SetRotate(maxSize, maxBackups, maxAge int) {
	var lumberjackLogger *lumberjack.Logger
	if logger.filename == "" {
		lumberjackLogger = &lumberjack.Logger{
			Filename:   defaultFilename,
			MaxSize:    maxSize,
			MaxBackups: maxBackups,
			MaxAge:     maxAge,
		}
		logger.filename = defaultFilename
		multiIO := io.MultiWriter(os.Stdout, lumberjackLogger)
		logger.SetOutput(multiIO)
		logger.lumberjackLogger = lumberjackLogger
	} else {
		lumberjackLogger = &lumberjack.Logger{
			Filename:   logger.filename,
			MaxSize:    maxSize,
			MaxBackups: maxBackups,
			MaxAge:     maxAge,
		}
		multiIO := io.MultiWriter(os.Stdout, lumberjackLogger)
		logger.SetOutput(multiIO)
		logger.lumberjackLogger = lumberjackLogger
	}
}

// SetLevel change log level
func SetLevel(l Level) {
	logger.SetLevel(transLevelToLogrusLevel(l))
}

// SetOutput sets the logger output.
// Be sure you want set your output, if you just want set logger
// output to a file, use SetOutputFile
func SetOutput(output io.Writer) {
	logger.SetOutput(output)
}

// Tracef see logrus Tracef
func Tracef(format string, args ...interface{}) {
	logger.Tracef(format, args...)
}

// Debugf see logrus Debugf
func Debugf(format string, args ...interface{}) {
	logger.Debugf(format, args...)
}

// Infof see logrus Infof
func Infof(format string, args ...interface{}) {
	logger.Infof(format, args...)
}

// Warnf see logrus Warnf
func Warnf(format string, args ...interface{}) {
	logger.Warnf(format, args...)
}

// Errorf see logrus Errorf
func Errorf(format string, args ...interface{}) {
	logger.Errorf(format, args...)
}

// Fatalf will exit(1), see logrus Fatalf
func Fatalf(format string, args ...interface{}) {
	logger.Fatalf(format, args...)
}

// Panicf will panic, see logrus Panicf
func Panicf(format string, args ...interface{}) {
	logger.Panicf(format, args...)
}
